sdk/src/main/scala/com/horizen/certificatesubmitter/strategies/WithKeyRotationCircuitStrategy.scala

```diff
@@ -0,0 +1,180 @@
+package com.horizen.certificatesubmitter.strategies
+
+import com.horizen.block.SidechainCreationVersions.SidechainCreationVersion
+import com.horizen.block.{SidechainBlockBase, SidechainBlockHeaderBase, WithdrawalEpochCertificate}
+import com.horizen.certificatesubmitter.AbstractCertificateSubmitter.SignaturesStatus
+import com.horizen.certificatesubmitter.dataproof.CertificateDataWithKeyRotation
+import com.horizen.certificatesubmitter.keys.{CertifiersKeys, KeyRotationProof, SchnorrKeysSignatures}
+import com.horizen.certnative.BackwardTransfer
+import com.horizen.chain.AbstractFeePaymentsInfo
+import com.horizen.cryptolibprovider.{CryptoLibProvider, ThresholdSignatureCircuitWithKeyRotation}
+import com.horizen.params.NetworkParams
+import com.horizen.proposition.SchnorrProposition
+import com.horizen.storage.AbstractHistoryStorage
+import com.horizen.transaction.Transaction
+import com.horizen._
+import sparkz.core.transaction.MemoryPool
+
+import java.util
+import java.util.Optional
+import scala.collection.JavaConverters._
+import scala.compat.java8.OptionConverters.RichOptionForJava8
+import scala.reflect.ClassTag
+import scala.util.Try
+
+class WithKeyRotationCircuitStrategy[
+  TX <: Transaction,
+  H <: SidechainBlockHeaderBase,
+  PM <: SidechainBlockBase[TX, H] : ClassTag,
+  _FPI <: AbstractFeePaymentsInfo,
+  _HSTOR <: AbstractHistoryStorage[PM, _FPI, _HSTOR],
+  _HIS <: AbstractHistory[TX, H, PM, _FPI, _HSTOR, _HIS],
+  _MS <: AbstractState[TX, H, PM, _MS],
+  _VL <: Wallet[SidechainTypes#SCS, SidechainTypes#SCP, TX, PM, _VL],
+  _MP <: MemoryPool[TX, _MP]](settings: SidechainSettings, params: NetworkParams,
```

Comment:
I have a feeling that here we have some mess with the generics and with the type definitions.
In case we decide to refactor it we need to start from `AbstractCertifcateSubmitter` and move `type` definitions to class templates.